ReactSkia/core_modules/RSkInputEventManager.cpp

```diff
@@ -26,40 +30,104 @@ RSkInputEventManager::RSkInputEventManager(){
   spatialNavigator_ =  SpatialNavigator::RSkSpatialNavigator::sharedSpatialNavigator();
   keyRepeat=false;
   previousKeyType=RNS_KEY_UnKnown;
+#if ENABLE(FEATURE_KEY_THROTTLING)
+  inputWorkerThread_ = std::thread(&RSkInputEventManager::inputWorkerThreadFunction, this);
+  sem_init(&keyEventPost_, 0, 1);
+  completeCallback_ = std::bind(&RSkInputEventManager::onEventComplete, this);
+#else
+  completeCallback_ = nullptr;
+#endif
+}
+
+RSkInputEventManager::~RSkInputEventManager(){
+  if(eventId_ > 0) {
+    NotificationCenter::defaultCenter().removeListener(eventId_);
+    eventId_ = 0;
+  }
+  if(subWindowEventId_ > 0) {
+    NotificationCenter::subWindowCenter().removeListener(subWindowEventId_);
+    subWindowEventId_ = 0;
+  }
+#if ENABLE(FEATURE_KEY_THROTTLING)
+  sem_destroy(&keyEventPost_);
+  if(keyQueue_)
+    keyQueue_ = nullptr;
+  if(inputWorkerThread_.joinable()) {
+    inputWorkerThread_.join();
+  }
+#endif
+}
+
+#if ENABLE(FEATURE_KEY_THROTTLING)
+void RSkInputEventManager::inputWorkerThreadFunction() {
+  RskKeyInput keyInput;
+  while(true) {
+    if(activeInputClients_ > 0) // If there are clients who are still processing previous key then wait..
+      sem_wait(&keyEventPost_);
+    keyQueue_->pop(keyInput); // Blocks if empty. TODO better to use timed pop to avoid unforseen block issues ??
+    RNS_LOG_DEBUG("Process input from queue,  Key : " << keyInput.key_ << " Action : " <<keyInput.action_ << ", Repeat : " <<keyInput.repeat_);
+    processKey(keyInput);
+  }
+}
+
+void RSkInputEventManager::onEventEmit() {
+  RNS_LOG_DEBUG("++++++++++ On Key Event Emit ++++++++++ : Pending Events : " << activeInputClients_.load());
+  activeInputClients_++;
 }
 
+void RSkInputEventManager::onEventComplete() {
+  (activeInputClients_ > 0) ? ((--activeInputClients_ == 0) ? sem_post(&keyEventPost_) : false) : false;
```

Comment:
Will there be a possibility : activeInputClients_ will be > 1 and possibly blocks the key processing

Reply:
The idea of key throttling is to block the key processing until previous key is completely processed by all the clients.